File: lib/php/tests/UsersTest.php
<?php

declare(strict_types=1);

namespace WgPortal\Client\Tests;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use PHPUnit\Framework\TestCase;
use WgPortal\Client\HttpClient;
use WgPortal\Client\UsersApi;
use WgPortal\Client\WgPortalException;

class UsersTest extends TestCase
{
    private const BASE_URL = 'http://wg-portal.test:8888';

    private function makeApi(array &$queue): UsersApi
    {
        $ref = new \ReflectionClass(HttpClient::class);
        $http = $ref->newInstanceWithoutConstructor();

        $prop = $ref->getProperty('baseUrl');
        $prop->setAccessible(true);
        $prop->setValue($http, self::BASE_URL . '/api/v1');

        $headersProp = $ref->getProperty('headers');
        $headersProp->setAccessible(true);
        $headersProp->setValue($http, ['Authorization' => 'Basic dGVzdDp0ZXN0']);

        $mock = new MockHandler($queue);
        $httpProp = $ref->getProperty('http');
        $httpProp->setAccessible(true);
        $httpProp->setValue($http, new GuzzleClient(['handler' => HandlerStack::create($mock)]));

        return new UsersApi($http);
    }

    private function fixtureUser(): array
    {
        return [
            'Identifier' => 'uid-1234567',
            'Email' => 'test@test.com',
            'Firstname' => 'Max',
            'Lastname' => 'Muster',
            'IsAdmin' => false,
            'ApiEnabled' => false,
            'PeerCount' => 2,
        ];
    }

    public function testGetAll(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);
        $queue[] = new Response(200, ['Content-Type' => 'application/json'], json_encode([
            $this->fixtureUser(),
            ['Identifier' => 'uid-admin', 'Email' => 'admin@wgportal.local', 'IsAdmin' => true],
        ]));

        $result = $api->getAll();
        $this->assertCount(2, $result);
        $this->assertEquals('uid-1234567', $result[0]['Identifier']);
    }

    public function testGetById(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);
        $queue[] = new Response(200, ['Content-Type' => 'application/json'], json_encode($this->fixtureUser()));

        $result = $api->getById('uid-1234567');
        $this->assertEquals('test@test.com', $result['Email']);
    }

    public function testCreate(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);
        $queue[] = new Response(200, ['Content-Type' => 'application/json'], json_encode($this->fixtureUser()));

        $result = $api->create($this->fixtureUser());
        $this->assertEquals('uid-1234567', $result['Identifier']);
    }

    public function testDelete(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);
        $queue[] = new Response(204);

        $api->delete('uid-1234567');
        $this->assertTrue(true);
    }

    public function testGetByIdThrowsOn404(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);
        $queue[] = new Response(404, ['Content-Type' => 'application/json'], '{"Code":404,"Message":"Not Found"}');

        $this->expectException(WgPortalException::class);
        $api->getById('nonexistent');
    }
}

File: lib/rust/tests/users_test.rs
use mockito::Server;
use serde_json::json;
use wg_portal_client::{WgUser, WgPortalClient};

fn setup() -> (mockito::ServerGuard, WgPortalClient) {
    let server = Server::new();
    let client = WgPortalClient::new(&server.url(), "admin@wgportal.local", "test-token");
    (server, client)
}

fn fixture_user() -> serde_json::Value {
    json!({
        "Identifier": "uid-1234567",
        "Email": "test@test.com",
        "Firstname": "Max",
        "Lastname": "Muster",
        "IsAdmin": false,
        "ApiEnabled": false,
        "PeerCount": 2
    })
}

fn fixture_user_list() -> serde_json::Value {
    json!([fixture_user(), {
        "Identifier": "uid-admin",
        "Email": "admin@wgportal.local",
        "IsAdmin": true,
        "ApiEnabled": true
    }])
}

#[test]
fn get_all() {
    let (mut server, client) = setup();
    let mock = server
        .mock("GET", "/api/v1/user/all")
        .with_body(fixture_user_list().to_string())
        .with_header("content-type", "application/json")
        .create();

    let result = client.users().get_all().unwrap();
    assert_eq!(result.len(), 2);
    assert_eq!(result[0].identifier, "uid-1234567");
    mock.assert();
}

#[test]
fn get_by_id() {
    let (mut server, client) = setup();
    let mock = server
        .mock("GET", "/api/v1/user/by-id/uid-1234567")
        .with_body(fixture_user().to_string())
        .with_header("content-type", "application/json")
        .create();

    let result = client.users().get_by_id("uid-1234567").unwrap();
    assert_eq!(result.email.as_deref(), Some("test@test.com"));
    mock.assert();
}

#[test]
fn create_user() {
    let (mut server, client) = setup();
    let mock = server
        .mock("POST", "/api/v1/user/new")
        .with_body(fixture_user().to_string())
        .with_header("content-type", "application/json")
        .create();

    let data: WgUser = serde_json::from_value(fixture_user()).unwrap();
    let result = client.users().create(&data).unwrap();
    assert_eq!(result.identifier, "uid-1234567");
    mock.assert();
}

#[test]
fn delete_user() {
    let (mut server, client) = setup();
    let mock = server
        .mock("DELETE", "/api/v1/user/by-id/uid-1234567")
        .with_status(204)
        .create();

    client.users().delete("uid-1234567").unwrap();
    mock.assert();
}
